src/test/scala/firrtlTests/MemLatencySpec.scala

```diff
@@ -0,0 +1,103 @@
+  val idle = Seq(Poke("m.w.en", 0), Poke("m.r.en", 0))
+  def pokes = (wPokes zip rPokes).map { case (wp, rp) => wp ++ rp } ++ Seq.fill(rLatency)(idle)
+
+  // Need to delay read value expects by <rLatency>
+  def expects = Seq.fill(rLatency)(Seq(Step(1))) ++ memAccesses.map {
+    case MemAccess(_, Some(Read(_, expected))) => Seq(Expect("m.r.data", expected), Step(1))
+    case _ => Seq(Step(1))
+  }
+
+  def commands: Seq[SimpleTestCommand] = (pokes zip expects).flatMap { case (p, e) => p ++ e }
+}
+
+trait ReadEnToggles {
+  import MemLatencySpec._
+  val memAccesses: Seq[MemAccess] = Seq(
+    MemAccess(Some(Write(6, 32)),  None),
+    MemAccess(Some(Write(14, 87)), None),
+    MemAccess(None,                None),
+    MemAccess(Some(Write(19, 63)), Some(Read(14, 87))),
+    MemAccess(Some(Write(22, 49)), None),
+    MemAccess(Some(Write(11, 99)), Some(Read(6, 32))),
+    MemAccess(Some(Write(42, 42)), None),
+    MemAccess(Some(Write(77, 81)), None),
+    MemAccess(Some(Write(13, 7)),  Some(Read(19, 63))),
+    MemAccess(Some(Write(39, 5)),  Some(Read(42, 42)))
+  )
+}
+
+/*
+ *  This framework is for execution tests, so these tests all focus on
+ *  *legal* configurations. Illegal memory parameters that should
+ *  result in errors should be tested in MemSpec.
+ */
+
+// These two are the same in practice, but the two tests could help expose bugs in VerilogMemDelays
+class CombMemSpecNewRUW extends MemLatencySpec(rLatency = 0, wLatency = 1, ruw = "new") with ReadEnToggles
+class CombMemSpecOldRUW extends MemLatencySpec(rLatency = 0, wLatency = 1, ruw = "old") with ReadEnToggles
+
+// Odd combination: combinational read with 2-cycle write latency
+class CombMemSpecNewRUWRL2 extends MemLatencySpec(rLatency = 0, wLatency = 2, ruw = "new") with ReadEnToggles
+class CombMemSpecOldRUWRL2 extends MemLatencySpec(rLatency = 0, wLatency = 2, ruw = "old") with ReadEnToggles
+
+// Standard sync read mem
+class WriteFirstMemReadEnToggles extends MemLatencySpec(rLatency = 1, wLatency = 1, ruw = "new") with ReadEnToggles
+class ReadFirstMemReadEnToggles extends MemLatencySpec(rLatency = 1, wLatency = 1, ruw = "old") with ReadEnToggles
+
+// Read latency 2
+class WriteFirstMemReadEnTogglesRL2 extends MemLatencySpec(rLatency = 2, wLatency = 1, ruw = "new") with ReadEnToggles
+class ReadFirstMemReadEnTogglesRL2 extends MemLatencySpec(rLatency = 2, wLatency = 1, ruw = "old") with ReadEnToggles
+
+// Write latency 2
+class WriteFirstMemReadEnTogglesWL2 extends MemLatencySpec(rLatency = 1, wLatency = 2, ruw = "new") with ReadEnToggles
+class ReadFirstMemReadEnTogglesWL2 extends MemLatencySpec(rLatency = 1, wLatency = 2, ruw = "old") with ReadEnToggles
+
+// Read latency 2, write latency 2
+class WriteFirstMemReadEnTogglesRL2WL2 extends MemLatencySpec(rLatency = 2, wLatency = 2, ruw = "new") with ReadEnToggles
+class ReadFirstMemReadEnTogglesRL2WL2 extends MemLatencySpec(rLatency = 2, wLatency = 2, ruw = "old") with ReadEnToggles
```

Comment:
This kind of test is very cool!
Nice job!

Reply:
Thanks